### ISSUE TITLE
Beta 17: Vital callback-based player login/logout methods missing (+more)

### ISSUE THREAD

**Issue by mhuusko5** [spotify/ios-sdk#697]
TLDR: Please re-expose callback based player login/logout methods, as they're vital to working around player SDK bugs. @erdtman @Voxar 

With the 200+ open issues and two-bug-fixes-(and-a-new-bug)-or-build-config-change-per-two-months deployment strategy, there's obviously a lot of work as a 3rd party dev that goes into making this SDK usable in production. Avoiding internal crashes during API requests with crazy swizzling/exception handling, checking SPT objects internal properties for error information when an SDK call succeeds but an object is deformed, periodic force session renewals when a session is reporting valid but the player fails to login anyway, a player -logout before every -login to prevent the sometimes-login-simply-doesnt-return issue, and of course working around the notorious network/LTE-switch which leaves all player operations not calling back, requiring an app restart (best solution being monitoring reachability, and logging out/then back in player and re-playing/setting state.. works 90% of the time, but of course causes the occasional crash).

One would have hoped those latter few player bugs might have been fixed in beta 17, but instead the serial/callback-based logout/login methods vital to the painful workarounds necessary for those bugs have been removed.

*** Please if possible, a patch release with those two methods re-exposed would be greatly appreciated.

Unrelated to this specific issue (but clearly I'm overloading this issue anyway for a general message), a little more transparency regarding the development of the SDK would be even more greatly appreciated (and I imagine I can speak on behalf of everyone who participates on this repo in that regard).
From an outside view, with important issues/tickets often not (or not properly, or not timely) responded to, with the main people who post releases/respond to issues seemingly changing every few months, and with app/product breaking issues (e.g. LTE!) left unfixed for months (actually we're dealing in the timeframe of years at this point) while unrequested features (manageable player thread lifetime??) are added.... well... it's just not very clear who this SDK is being developed for...

*** ... if it's for the people testing/using the SDK and diligently reporting issues/participating in discussion on this repo (and providing peer support/advice when official support/advice isn't present), then perhaps some adjusts to strategy are due. Starting with, perhaps, a group bug prioritisation session?

*** ... if it's somehow not for this general public, but instead some other party/stakeholder with differing specific feature/bug priorities, okay, but let us know, so we can prioritise our development efforts on our products accordingly.

*** ... most importantly, if the SDK simply isn't in active development anymore/doesn't have a dedicated team anymore (which is one suspicion I've had/inference that could be made from the disparate release schedule), and is basically in maintenance mode, *please please please let us know* – because we are developing products and basing business decisions around this SDK one day working/being production quality, and if it's not, we need to know so we can pursue alternatives.

As a lover of the Spotify product/brand, a fellow developer, and a knower-of-how-difficult architecting/maintaining large media frameworks is.. many thanks for the work done so far, and I look forward to thoughts/information that can help this community bring the SDK out of beta in a timely manner!

**Comment by NOlivNeto** [comment]
Can't +1 this enough.

**Comment by erdtman** [comment]
Sorry for the delay.
I might be missing something but why cant you use the SPTAudioStreamingDelegate and the callbacks
-(void)audioStreamingDidLogin:(SPTAudioStreamingController *)audioStreaming;
-(void)audioStreamingDidLogout:(SPTAudioStreamingController *)audioStreaming;
instead of the block callback.

**Comment by mhuusko5** [comment]
Those aren't callbacks, those are delegate methods. Callbacks capture state, delegate methods do not (unless providing an additional 'context' or similar field).
Once control flow reaches `-audioStreamingDidLogin:`, theres no context to know whether it was called just because, whether it was called because I previously wanted to play something, or whether it was called because I'm trying to do a logout then login then resume to fix LTE bug, etc.
This is pretty simple stuff. Delegate methods don't allow for sequential asynchronous programming, that's why they started to fall out of favour.. 7 years ago.

Anyway, went ahead a created a rudimentary callback based system for login/logout by storing an array of callbacks for each, and iterating them/calling them successfully in `..DidLogin:/Logout:`, or calling them with an error in `didEncounterError:` if the error seems related. Heinous, but seems to work.
